### PR TITLE
Document and enhance QTT.

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 
I thought for the benefit of our new starters and the community I'd improve the documentation and functionality of QTT a.k.a. `QueryTranslationTest`.

Main changes:
- key in input and output messages is now optional
- timestamp in input and output messages is now optional
- inputs and outputs no longer required if expectedException supplied
- statements can now contain statements that don't parse. (the error will be matched against expectedExcetion).
- Documentation added!
- README is linked to from our CONTRIBUTING page.

### Testing done 
No production code changed.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/test/resources/query-validation-tests/README.md [hunk 1]
```diff
@@ -5,9 +5,191 @@ file contains one or more tests. They are written in a simple json format that d
...
+        {"topic": "input_topic", "key": 1, "value": "{"id": 9}", "timestamp": 30000},
+        {"topic": "input_topic", "key": 1, "value": "{"id": 11}", "timestamp": 40000}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": 0, "value": "{"id": 8}", "timestamp": 0},
+        {"topic": "OUTPUT", "key": 0, "value": "{"id": 9}", "timestamp": 30000}
+     ]
+    },
+    {
+      "name": "my first negative test",
```
Comment (agavra) [inline ksql-engine/src/test/resources/query-validation-tests/README.md]: tbh, I'm not sure I agree that this is the right place for negative tests - it seems kind-of heavy to setup an end-to-end test just to test something that doesn't parse. I would be more inclined to create a simple parser unit test util that allows us to unit test negative examples

File: ksql-engine/src/test/resources/query-validation-tests/README.md [hunk 2]
```diff
@@ -5,9 +5,191 @@ file contains one or more tests. They are written in a simple json format that d
...
+### Topics
+A test case can optionally supply an array of topics that should exist and additional information about those topics.
+It is not necessary to add entries for topics required by the test case unless any of the following attributes need to be controlled:
+(Tests will create input and output topics as required).
+
+| Attribute | Description |
+|-----------|:------------|
+| name      | (Required) the name of the topic |
+| schema    | (Optional) the schema, registered in the schema store, of the topic. |
+| format    | (Optional) the serialization format of records within the topic, e.g. `AVRO` |
```
Comment (agavra) [inline ksql-engine/src/test/resources/query-validation-tests/README.md]: are there defaults for these Optional values?

File: ksql-engine/src/test/resources/query-validation-tests/README.md [hunk 3]
```diff
@@ -5,9 +5,191 @@ file contains one or more tests. They are written in a simple json format that d
...
+  ]
+}
+
+```
+
+Test cases should be written to be as succinct as possible to make them as readable as possible.
+For example:
+ - the schemas in DDL statements should be as minimal as possible. So avoid cutting & pasting large schemas between test cases.
+ - keep the test of source messages to the minimum required to prove the functionality and any edge cases.
+ - keep the definition of source and sink messages succinct too: don't add optional attributes if they are not pertinant to the test case.
```
Comment (blueedgenick) [inline ksql-engine/src/test/resources/query-validation-tests/README.md]: small typo: pertinant -> pertinent

File: ksql-engine/src/test/resources/query-validation-tests/README.md [hunk 4]
```diff
@@ -5,9 +5,191 @@ file contains one or more tests. They are written in a simple json format that d
 the inputs, and the expected outputs. These queries will be translated by KSQL into a Streams
 topology, executed, and verified.
 
-You can specify multiple statements per test case, i.e., to set up the various streams needed,
-for joins etc, but currently only the final topology will be verified. This should be enough,
+The test cases are run by the `QueryTranslationTest` test class.
+
+## Topology comparision
+These test also validate the generated topology matches the expected topology,
+i.e. a test will fail if the topology has changed from previouus runs.
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/test/resources/query-validation-tests/README.md]: done

File: ksql-engine/src/test/resources/query-validation-tests/README.md (hunk 1 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/resources/query-validation-tests/README.md]: I can see your point of view, but I feel that allowing negative tests in these JSON functional tests means we're more likely to see contributions testing that users are seeing sensible error messages for unsupported operations.  

I'm not particularly talking about the error a user will see if they mistype something. (I agree this belongs in a unit test). I'm talking more about statements that a user might think are valid, but aren't supported, e.g. GROUP BY functionality where the projection contains columns that aren't an exact match for the group by clause.  Such testing of what is _not_ supported, IMHO, belongs with the functional testing of what _is_ supported.

File: ksql-engine/src/test/resources/query-validation-tests/README.md (hunk 2 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/resources/query-validation-tests/README.md]: updated.

File: ksql-engine/src/test/resources/query-validation-tests/README.md (hunk 3 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/resources/query-validation-tests/README.md]: Thanks. Done.
